peer/pendingheap/heap.go

```diff
@@ -0,0 +1,186 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package pendingheap
+
+import (
+	"container/heap"
+	"context"
+	"fmt"
+	"sync"
+
+	"go.uber.org/yarpc/api/peer"
+	"go.uber.org/yarpc/api/transport"
+)
+
+type pendingHeap struct {
+	sync.RWMutex
+
+	transport peer.Transport
+	peers     []*peerScore
+
+	// next is an incrementing counter for every push, which is compared when
+	// scores are equal. This ends up implementing round-robin when scores are
+	// equal.
+	next int
+}
+
+func (ph *pendingHeap) Choose(ctx context.Context, req *transport.Request) peer.StatusPeer {
+	ph.Lock()
+	ps, ok := ph.popPeer()
+	if !ok {
+		ph.Unlock()
+		return nil
+	}
+
+	// Note: We push the peer back to reset the "next" counter.
+	// This gives us round-robin behavior.
+	ph.pushPeer(ps)
+
+	ph.Unlock()
+	return ps.peer
+}
+
+func (ph *pendingHeap) Add(p peer.StatusPeer) peer.Subscriber {
+	if p == nil {
+		return nil
+	}
+
+	ps := &peerScore{peer: p, heap: ph}
+	ps.score = scorePeer(p)
+
+	ph.Lock()
+	ph.pushPeer(ps)
+	ph.Unlock()
+	return ps
+}
+
+func (ph *pendingHeap) Remove(p peer.StatusPeer, sub peer.Subscriber) {
+	ps, ok := sub.(*peerScore)
+	if !ok {
+		return
+	}
+
+	ph.Lock()
+	ph.delete(ps.idx)
+	ph.Unlock()
+}
+
+func (ph *pendingHeap) notifyStatusChanged(ps *peerScore) {
+	status := ps.peer.Status()
+	ph.Lock()
+	ps.status = status
+	ps.score = scorePeer(ps.peer)
+	ph.update(ps.idx)
+	ph.Unlock()
+}
+
+func (ph *pendingHeap) Len() int {
+	l := len(ph.peers)
+	return l
+}
+
+// Less returns whether the left peer has a lower score. If the scores are
+// equal, it returns the older peer (where "last" is lower.)
+func (ph *pendingHeap) Less(i, j int) bool {
+	p1 := ph.peers[i]
+	p2 := ph.peers[j]
+	if p1.score == p2.score {
+		return p1.last < p2.last
+	}
+	return p1.score < p2.score
+}
+
+// Swap implements the heap.Interface. Do NOT use this method directly.
+func (ph *pendingHeap) Swap(i, j int) {
+	p1 := ph.peers[i]
+	p2 := ph.peers[j]
+
+	ph.peers[i], ph.peers[j] = ph.peers[j], ph.peers[i]
+	p1.idx = j
+	p2.idx = i
+}
+
+// Push implements the heap.Interface. Do NOT use this method directly.
+// Use pushPeer instead.
+func (ph *pendingHeap) Push(x interface{}) {
+	ps := x.(*peerScore)
+	ps.idx = len(ph.peers)
+	ph.peers = append(ph.peers, ps)
+}
+
+// Pop implements the heap.Interface. Do NOT use this method directly.
+// Use popPeer instead.
+func (ph *pendingHeap) Pop() interface{} {
+	lastIdx := len(ph.peers) - 1
+	last := ph.peers[lastIdx]
+	ph.peers = ph.peers[:lastIdx]
+	return last
+}
+
+func (ph *pendingHeap) delete(idx int) {
+	// Swap the element we want to delete with the last element, then pop it off.
+	ph.Swap(idx, ph.Len()-1)
+	ph.Pop()
+
+	// If the original index still exists in the list, it contains a different
+	// element so update the heap.
+	if idx < ph.Len() {
+		ph.update(idx)
+	}
+}
+
+func (ph *pendingHeap) validate(ps *peerScore) error {
```

Comment:
could this function exist in the _test file?

Reply:
✔️